File: stack/php/src/app/Stack.php
<?php
declare(strict_types=1);

class Stack
{
    private int $capacity = 0;
    
    private array $values = [];

    /**
     * Allocates enough memory for a required capacity.
     */
    public function allocate(int $capacity): void
    {
        $this->capacity = $capacity;
    }

    /**
     * Returns the current capacity.
     */
    public function getCapacity(): int
    {
        return $this->capacity;
    }

    /**
     * Clear stack.
     */
    public function clear(): void
    {
        $this->values = [];
    }

    /**
     * Returns a shallow copy of the stack.
     */
    public function copy(): Stack
    {
        return clone $this;
    }

    /**
     * Returns whether the stack is empty.
     */
    public function isEmpty(): bool
    {
        return count($this->values) < 1;
    }

    /**
     * Returns the value at the front of the stack.
     */
    public function peek(): int
    {
        return $this->values[0];
    }

    /**
     * Removes and returns the value.
     */
    public function pop(): int
    {
        return array_shift($this->values);
    }

    /**
     * Pushes values into the stack.
     */
    public function push(int $value): void
    {
        if ($this->capacity === count($this->values)) {
            throw new InvalidArgumentException(
                'Not enough memory. Capacity: ' . $this->capacity
            );
        }

        $values = array_reverse($this->values);
        $values[] = $value;
        $this->values = array_reverse($values);
    }

    /**
     * Converts the queue to an array.
     */
    public function toArray(): array
    {
        return $this->values;
    }
}

File: stack/go/app/stack.go
package main

import (
	"fmt"
	"log"
)

type Stack struct {
	capacity int
	values   []int
}

func (s *Stack) Allocate(capacity int) {
	s.capacity = capacity
}

func (s Stack) GetCapacity() int {
	return s.capacity
}

func (s *Stack) Clear() {
	s.values = nil
}

func (s Stack) IsEmpty() bool {
	return len(s.values) < 1
}

func (s Stack) Copy() Stack {
	clone := s

	return clone
}

func (s *Stack) Push(value int) {
	s.validateCapacity()
	s.values = append(s.values, 0)
	copy(s.values[1:], s.values)
	s.values[0] = value
}

func (s *Stack) Peek() int {
	return s.values[0]
}

func (s *Stack) Pop() int {
	var value int

	value, s.values = s.values[0], s.values[1:]

	return value
}

func (s Stack) validateCapacity() {
	if len(s.values) > s.capacity {
		log.Fatal(fmt.Sprintf("Not enough memory. Capacity: %d", s.capacity))
	}
}
